extensions/src/main/java/org/pentaho/platform/plugin/services/pluginmgr/PluginMessages.java

```diff
@@ -0,0 +1,91 @@
+package org.pentaho.platform.plugin.services.pluginmgr;
+
+import org.pentaho.platform.util.StringUtil;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.ResourceBundle;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class PluginMessages {
+  private final Pattern KEY_PATTERN = Pattern.compile( "\\$\\{([^}]+)\\}" );
+
+  private final Map<Locale, ResourceBundle> locales = Collections
+    .synchronizedMap( new HashMap<Locale, ResourceBundle>() );
+
+  private final String bundleName;
+
+  private final ClassLoader bundleLoader;
+
+  public PluginMessages( String bundleName, PluginClassLoader bundleLoader ) {
+    this.bundleName = bundleName;
+    this.bundleLoader = bundleLoader;
+  }
+
+  private ResourceBundle getBundle( Locale locale ) {
+    ResourceBundle bundle = locales.get( locale );
+
+    if ( bundle == null ) {
+      bundle = ResourceBundle.getBundle( bundleName, locale, bundleLoader );
+      locales.put( locale, bundle );
+    }
+
+    return bundle;
+  }
+
+  /**
+   * Returns a localized string for the given key and locale.
+   * If the key is not found, the original value is returned.
+   *
+   * @param value the value string to be localized
+   * @param locale the locale to use for localization
+   * @return the localized string or the original value if not found
+   */
+  public String getString( String value, Locale locale ) {
```

Comment:
This logic could be split by directly receiving the `ResourceBundle` on which to evaluate the string value.

Reply:
changed the logic as discussed offline